### PR TITLE
refactor: Settings code cleanups

### PR DESCRIPTION
**Pull request by ryanofsky**
This PR doesn't change behavior. It just implements some suggestions from #15934 and #16545 and few other small cleanups.

### CODE REVIEW

**Review by hebasto — COMMENTED** [review]
Concept ACK 6f018d367302fdf79174c0981e76fc32ec9d77b2

File: src/util/system.h [hunk 1]
```diff
@@ -270,6 +270,27 @@ class ArgsManager
      */
     std::string GetChainName() const;
 
+    /**
+     * Returns true if settings values from the default section should be used,
+     * depending on the current network and whether the setting is
+     * network-specific.
+     */
+    bool UseDefaultSection(const std::string& arg) const EXCLUSIVE_LOCKS_REQUIRED(cs_args);
```
Comment (promag) [inline src/util/system.h]: nit, could be private.

**Review by promag — COMMENTED** [review]
Concept ACK. Is bcf8060557d544fa881be9e74c2e108189010dca really necessary, not sure the motivation of it.

**Comment by ryanofsky** [comment]
> Concept ACK. Is [bcf8060](https://github.com/bitcoin/bitcoin/commit/bcf8060557d544fa881be9e74c2e108189010dca) really necessary, not sure the motivation of it.

As the commit message says, some of the flags are incorrectly applied and cause tests to fail when type checking is actually implemented in #16545. At first I tried to selectively update the flags instead of using a scripted diff, but as I was iterating on #16545 more of these old tests kept failing and I kept going back and forth on the flags. I think bottom line is these tests were never written to interact with flags, and I have much better dedicated tests for the flags in #16545. I doubt #16545 will be the last PR changing these flags, and I want to avoid old tests becoming a maintenance burden for more flag changes in the future.

Reason for including the commit in this PR instead of delaying it for the future, is that right now the tests are misleading and make it look like unused unimplemented constants could actually have an effect.

**Review by jnewbery — COMMENTED** [review]
utACK 9926e06fe9aa258ae06bf2b6e9c7f589f7c8a565 with just a few small style nits

File: src/test/util_tests.cpp [hunk 2]
```diff
@@ -189,12 +190,119 @@ struct TestArgsManager : public ArgsManager
...
+    CheckValue(M::ALLOW_ANY, "-novalue=0", Expect{true}.String("1").Int(1).Bool(true).List({"1"}));
+    CheckValue(M::ALLOW_ANY, "-novalue=1", Expect{false}.String("0").Int(0).Bool(false).List({}));
+    CheckValue(M::ALLOW_ANY, "-novalue=2", Expect{false}.String("0").Int(0).Bool(false).List({}));
+    CheckValue(M::ALLOW_ANY, "-novalue=abc", Expect{true}.String("1").Int(1).Bool(true).List({"1"}));
+    CheckValue(M::ALLOW_ANY, "-value", Expect{""}.String("").Int(0).Bool(true).List({""}));
+    CheckValue(M::ALLOW_ANY, "-value=", Expect{""}.String("").Int(0).Bool(true).List({""}));
+    CheckValue(M::ALLOW_ANY, "-value=0", Expect{"0"}.String("0").Int(0).Bool(false).List({"0"}));
+    CheckValue(M::ALLOW_ANY, "-value=1", Expect{"1"}.String("1").Int(1).Bool(true).List({"1"}));
+    CheckValue(M::ALLOW_ANY, "-value=2", Expect{"2"}.String("2").Int(2).Bool(true).List({"2"}));
+    CheckValue(M::ALLOW_ANY, "-value=abc", Expect{"abc"}.String("abc").Int(0).Bool(false).List({"abc"}));
```
Comment (ryanofsky) [inline src/test/util_tests.cpp]: In commit "refactor: Add util_CheckValue test" (425bb307252cf4dec9b3ef6426e6548b2be7a303)

Note: this test is extended in 909bcca7971063f42de6d981111c628bf5a0ebfa from #16545 to check other flags besides ALLOW_ANY

**Review by ryanofsky — COMMENTED** [review]
Updated 9926e06fe9aa258ae06bf2b6e9c7f589f7c8a565 -> e9fd366044e271632dc0e4f96e1c14f8e87213ae ([`pr/cleanset.2`](https://github.com/ryanofsky/bitcoin/commits/pr/cleanset.2) -> [`pr/cleanset.3`](https://github.com/ryanofsky/bitcoin/commits/pr/cleanset.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cleanset.2..pr/cleanset.3)) with suggestions from João and John

**Comment by jnewbery** [comment]
Code review ACK e9fd366044e271632dc0e4f96e1c14f8e87213ae

**Comment by MarcoFalke** [comment]
ACK e9fd366044 🚟

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e9fd366044 🚟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjBJQwAgNgNoo3sHg6IyTw00ye/IHSr222mR3aGg6NhiAoVziJ7US66yyt/c85I
s6nxdJe/5b4rkmNNID7bAh/VJNYAz4gNgMARcYngShCbQTRADKzeauIY8ZzBZFgm
W+uay6TWLKQM55E7y5jBBYmUFy1D5efmLtKR9wktp22eeMXlJMhLkI40nSfq4Tkt
aOSX6K7a5aaIImJs1SShl0nHsqF6z+al2MVuWpBxPZ34Wkn257p6G8BXwX6mNTNH
lxzjuPj6aHuWI496G7zbNBA6IkSGZgBE4/HcPCyYz+i16LzaqyYX/pWwtoRnSjME
BVCeqQEg3JFoNZtWsazzGpsh/Ki29X15jZwz1UrNYIXlf2v4QCC4Cj0CZe9P5QWc
+Ib+OyLoecz3+/WbGBWO6Lv+LKnPpyMjNj7bRNzJHikq6nw1wklbPMnOfxrJNbQv
xNTgZDLj6Z+Wp9TLD1jua7q2ciM+g7v3ADDYIh4UPJnLulAP8ML/pNKb0t1p5gne
L8iVPKtM
=p2lQ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `31d53ff8d0571f29f549510cc92e83930bbab33e7884245727870371ceabaaad  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010831d53ff8d0571f29f549510cc92e83930bbab33e7884245727870371ceabaaadf010e811694549760425a5138e426fd75ced08fff010487507f70f211dfac0f6a320be448d6108f1045dfbd7e4f008fb93eace897b1c820083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01074261cfa2d5d6261a074a53de4c5dc0c08f1045dfbd7e4f00897a34145d9bbfdc40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010cf1e2f2a9d3638caaea0f9af5f1722ff08f1045dfbd7e5f008b1ca87e37edc915b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d054cc61163b8b926a10822790fb805808f1045dfbd7e4f008e942a2a977be08e30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/test/getarg_tests.cpp [hunk 3]
```diff
@@ -43,7 +43,7 @@ static void SetupArgs(const std::vector<std::pair<std::string, unsigned int>>& a
 
 BOOST_AUTO_TEST_CASE(boolarg)
 {
-    const auto foo = std::make_pair("-foo", ArgsManager::ALLOW_BOOL);
+    const auto foo = std::make_pair("-foo", ArgsManager::ALLOW_ANY);
```
Comment (Evgeniy-boop) [inline src/test/getarg_tests.cpp]: ALLOW_ANY

File: src/test/util_tests.cpp [hunk 4]
```diff
@@ -189,12 +190,119 @@ struct TestArgsManager : public ArgsManager
...
+        if (expect.setting.isNull() || expect.setting.isFalse()) {
+            BOOST_CHECK_EQUAL(settings_list.size(), 0);
+        } else {
+            BOOST_CHECK_EQUAL(settings_list.size(), 1);
+            BOOST_CHECK_EQUAL(settings_list[0].write(), expect.setting.write());
+        }
+
+        if (expect.error) {
+            BOOST_CHECK(!success);
+            BOOST_CHECK_NE(error.find(expect.error), std::string::npos);
```
Comment (MarcoFalke) [inline src/test/util_tests.cpp]: Why not 

```diff
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index 86ea56ff36..f2f8ebd59a 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -243,7 +243,7 @@ public:
 
         if (expect.error) {
             BOOST_CHECK(!success);
-            BOOST_CHECK_NE(error.find(expect.error), std::string::npos);
+            BOOST_CHECK_EQUAL(error, expect.error);
         } else {
             BOOST_CHECK(success);
             BOOST_CHECK_EQUAL(error, "");
```

? I presume later tests require the less strict check?

Comment (ryanofsky) [reply src/test/util_tests.cpp]: I think it should be possible to make the change you're suggesting, which would give more helpful debug info if the check fails. This was originally written as part of #16545 which at one point was doing partial checks for error messages, but now I think checks the whole error string.

**Pull request merged**